core/trino-main/src/main/java/io/trino/sql/planner/optimizations/BeginTableWrite.java

```diff
@@ -176,10 +176,20 @@ public WriterTarget getTarget(PlanNode node)
                 return ((TableWriterNode) node).getTarget();
             }
             if (node instanceof DeleteNode) {
-                return ((DeleteNode) node).getTarget();
+                DeleteNode deleteNode = (DeleteNode) node;
+                DeleteTarget delete = deleteNode.getTarget();
+                return new DeleteTarget(
+                        findTableScanHandle(deleteNode.getSource()),
```

Comment:
We should also fix here
https://github.com/trinodb/trino/blob/075276b7b63dff9d68ce0afa681cf833b9cbf205/core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java#L480

-- the TH stored there is unused until here, and having it error-prone, it is IMO the sole reason we have the bugs being fixed here

Reply:
@martint and I had the same thought. I'll see how feasible it is to change this.

Reply:
Maybe `DeleteTarget.tableHandle` can be made Optional (sounds like taking a shortcut, i know ...)

Reply:
I changed it to optional.